### ISSUE TITLE
Strings with 0 preprended do not work

### ISSUE THREAD

**Issue by tosyu** [Samsung/iotjs#1062]
Hello,

I have a problem with strings that have 0 in their body. The problem could be problematic on low level communication (SPI or I2C) and enforces checking every byte prior to string concatenation.

The following code

```javascript
var str = 'hello world';
var newStr = String.fromCharCode(0) + str;
console.log(newStr);
```

will output

```bash
hello world
```

on nodejs

and in jerry script repl mode will output

```bash
\u0000hello world
```

but on iotjs the output is empty :cry:

**Comment by LaszloLango** [comment]
The problem comes from the implementation of the console module.

```c
fprintf(out_fd, "%s", iotjs_string_data(&msg));
```

It expects a zero terminated string, so it stops at the first zero character. To fix this you can change it to 

```c
fprintf(out_fd, "%.*s", iotjs_string_data(&msg), iotjs_string_size(&msg));
```

**Comment by tosyu** [comment]
This does not resolve the issue as bash (or even zsh) terminate the string after \0, check the following code

```c
#include <stdio.h>

int main(int argc, char** argv) {
    char msg[] = "te\0st";

    fprintf(stdout, "%.*s", 5, msg);

    return 0;
}
```

which outputs

```bash
te
```

the \0 is not converted to an unicode character, which differs from nodejs and jerry-script repl console

**Comment by zherczeg** [comment]
I don't understand how console.log affects SPI or I2C.

**Comment by LaszloLango** [comment]
@tosyu in that case `console.log` must be implemented like the default print handler in JerryScript: https://github.com/jerryscript-project/jerryscript/blob/master/jerry-ext/handler/handler-print.c

Although this should not affect SPI and I2C.

**Comment by tosyu** [comment]
@zherczeg @LaszloLango  At the first post I did not say console affects SPI or I2C, as I thought the problem was in the String implementation itself. Guys, I'm submitting an issue, not the solution for it, I do not have to know why the error occurs and where it's actually is :)

Side note. I have discovered the error while testing SPI<->Arduino communication, because not waiting for the device to respond to the transfer causes the respond data to be prepended by 0, as SPDR register in Arduino is empty till the device writes something :wink: And I was dumping the data with console.log, which printed nothing.

**Comment by LaszloLango** [comment]
@tosyu it is still not clear for me what is wrong with the string implementation. JS strings have to be able to store null characters. :)

**Comment by zherczeg** [comment]
Yes, even IoT.js strings can store 0 character, that is why the size is also stored for strings (so you can check that a \0 is a terminator or intermediate character).

**Comment by tosyu** [comment]
@LaszloLango @zherczeg I really do not understand where is the discussion going :confused: I created an issue that the \0 in strings works differently with nodejs and iotjs and jerry-script.

**Comment by zherczeg** [comment]
I suspect the problem is in string print, since it does not print further characters after \0 (this is probably correct). You can fix it by printing characters in a loop, if console.log is the only problem.

**Comment by akosthekiss** [comment]
I think we should mark this issue as a bug or a feature request. @tosyu has explicitly stated that this is a problem for him, but he is not planning to try and fix it. @zherczeg, @LaszloLango Now it's for us, for the project team, to decide first whether it is really a misbehaviour (-> mark it as bug) or whether we are happy with the current way of working (-> mark it as feature request, and decide whether we are willing to change). And then we shall implement the code change as well, if any, as our priorities allow.

**Comment by kolipka** [comment]
fix applied in #1070
